compiler/modules/front.syntax/tests/syntaxNode.bal

```diff
@@ -0,0 +1,535 @@
+        return syntaxNodeFromErrorTypeDesc(td);
+    }
+    else if td is XmlSequenceTypeDesc {
+        return syntaxNodeFromXmlSequenceTypeDesc(td);
+    }
+    else if td is TableTypeDesc {
+        return syntaxNodeFromTableTypeDesc(td);
+    }
+    else {
+        return syntaxNodeFromUnaryTypeDesc(td);
+    }
+}
+
+function syntaxNodeFromTupleTypeDesc(TupleTypeDesc td) returns NonTerminalSyntaxNode {
+    SyntaxNode[] memberNodes = flattenSyntaxNodeList(from int i in 0 ..< td.members.length() select commaSeperatedSyntaxNode(i, syntaxNodeFromTypeDesc(td.members[i])));
+    TypeDesc? rest = td.rest;
+    return nonTerminalSyntaxNode(td, { token: "[", pos: td.startPos},
+                                     memberNodes,
+                                     rest != () ? [syntaxNodeFromTypeDesc(rest), { token: "..." }] : (),
+                                     { token: "]", pos: td.endPos});
+}
+
+function syntaxNodeFromArrayTypeDesc(ArrayTypeDesc td) returns NonTerminalSyntaxNode {
+    SyntaxNode[][] dimensions = from SimpleConstExpr? dimension in td.dimensions select dimension == () ? [{ token: "[" }, { token: "]" }]:
+                                                                                                          [{ token: "[" }, syntaxNodeFromExpr(dimension), { token: "]" }];
+    return nonTerminalSyntaxNode(td, syntaxNodeFromTypeDesc(td.member),
+                                     ...dimensions);
+}
+
+function syntaxNodeFromMappingTypeDesc(MappingTypeDesc td) returns NonTerminalSyntaxNode {
+    TypeDesc|INCLUSIVE_RECORD_TYPE_DESC? rest = td.rest;
+    if td.fields.length() > 0 {
+        SyntaxNode[] fields = from FieldDesc f in td.fields select syntaxNodeFromFieldDesc(f);
+        return nonTerminalSyntaxNode(td, { token: "record", pos: td.startPos },
+                                         rest == INCLUSIVE_RECORD_TYPE_DESC ? { token: "{|" } : { token: "{" },
+                                         // JBUG: can't use query expression directly
+                                         fields,
+                                         (rest is TypeDesc) ? [syntaxNodeFromTypeDesc(rest), { token: "..." }, { token: ";" }] : (),
+                                         (rest is INCLUSIVE_RECORD_TYPE_DESC) ? { token: "}", pos: td.endPos } : { token: "|}", pos: td.endPos});
+    }
+    else {
+        return nonTerminalSyntaxNode(td, { token: "map", pos: td.startPos},
+                                         { token: "<" },
+                                         rest is TypeDesc ? syntaxNodeFromTypeDesc(rest) : { token: "never" },
+                                         { token: ">", pos: td.endPos });
+    }
+}
+
+function syntaxNodeFromFunctionTypeDesc(FunctionTypeDesc td, boolean functionSignature = false) returns SyntaxNode {
+    SyntaxNode[] params = flattenSyntaxNodeList(from int i in 0 ..< td.params.length() select commaSeperatedSyntaxNode(i, syntaxNodeFromFunctionTypeParam(td.params[i])));
```

Comment:
Don't flatten. There should be a syntax node for each param.

Reply:
Each param has a seperate node. What I am trying to do here is to flatten [[{ p }], [{ , }, { p }], [{ , }, { p }]] ({ p } is the syntax node for param and { , } is the syntax node for comma) it in to [{ p }, { , }, { p }, { , }, { p }]